janusgraph-test/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java

```diff
@@ -4023,6 +4021,8 @@ public void testTinkerPopOptimizationStrategies() {
         assertNumStep(1, 1, gts.V().has("id", numV / 5).has("weight", (numV / 5) % 5).as("x"), JanusGraphStep.class);
 
 
+        assertNumStep(superV * (numV / 5 * 2), 2, gts.V().has("id", sid).outE("knows").has("weight", P.between(1, 3)).inV().has("weight", P.between(1, 3)), JanusGraphStep.class, JanusGraphVertexStep.class);
```

Comment:
How can these tests verify your changes? I just removed all code changes and only kept these tests here to verify that they fail without your changes but the test method is still passing.
Unfortunately, I don't know how these changes could be tested else. Would it maybe be possible to add tests where a traversal pre-optimization is given and the optimized version is asserted [like it's done in `JanusGraphStepStrategyTest`](https://github.com/JanusGraph/janusgraph/blob/2fb971dd69abcd49e44dea33ccbe0d54a5df5357/janusgraph-test/src/test/java/org/janusgraph/blueprints/process/traversal/strategy/optimization/JanusGraphStepStrategyTest.java#L132)?

Reply:
I too was unsure how best to test this given that if it's working then it should return exactly the same results but faster. So we'd not expect the tests to fail without the changes, just to be slower when there are a large number of edges.

I verified that the two tests call the new code both by printing out temporary debug to stdout and by placing breakpoints in eclipse and stopping the debugger in the new methods.

I could/should look to add a check for the flavour of optimisation which uses the new  `JanusGraphEdgeVertexStep` class I think the profiler would pick that up. But the profile for the other flavour won't look any different.

Reply:
There's now a check to ensure that the profiling shows that `multiPreFetch=true` meaning the optimisation did kick in.